### PR TITLE
SONARPY-183 Include docstrings in the comment lines

### PR DESCRIPTION
**Pull request by ivandalbosco**

### CODE REVIEW

File: python-squid/src/main/java/org/sonar/python/metrics/FileLinesVisitor.java [hunk 1]
```diff
@@ -99,8 +112,36 @@ public void leaveFile(AstNode astNode) {
...
+        for (int line = location.startLine(); line <= location.endLine(); line++) {
+          linesOfDocstring.add(line);
+        }
+      }
+    }
+
+    return linesOfDocstring;
+  }
+
+  private void correctLinesOfCodeAndLineOfComments(Set<Integer> linesOfCode, Set<Integer> linesOfComments, Set<Integer> linesOfDosctring) {
```
Comment (SonarTech) [inline python-squid/src/main/java/org/sonar/python/metrics/FileLinesVisitor.java]: MAJOR Make "correctLinesOfCodeAndLineOfComments" a "static" method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

File: python-squid/src/main/java/org/sonar/python/metrics/FileLinesVisitor.java [hunk 2]
```diff
@@ -99,8 +112,36 @@ public void leaveFile(AstNode astNode) {
...
+    for (Integer line : linesOfDosctring) {
+      linesOfCode.remove(line);
+      linesOfComments.add(line);
+      System.out.println("linesOfComments = " + linesOfComments);
```
Comment (SonarTech) [inline python-squid/src/main/java/org/sonar/python/metrics/FileLinesVisitor.java]: MAJOR Replace this usage of System.out or System.err by a logger. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS106)

File: python-squid/src/main/java/org/sonar/python/DocstringVisitor.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+import org.sonar.python.api.PythonTokenType;
+import org.sonar.squidbridge.SquidAstVisitor;
+
+/**
+ * Visitor that retrieves all docstrings from a Python file.
+ * <p>
+ * Reminder: a docstring is a string literal that occurs as the first statement
+ * in a module, function, class, or method definition.
+ */
+public class DocstringVisitor<G extends Grammar> extends SquidAstVisitor<G> implements AstVisitor {
```
Comment (SonarTech) [inline python-squid/src/main/java/org/sonar/python/DocstringVisitor.java]: MINOR "AstVisitor" is implemented by a super class; there is no need to implement it here. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1939)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* MAJOR 3 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [PythonAstScanner.java#L100](https://github.com/SonarSource/sonar-python/blob/03e69be428808f500e5a4740cf3019a28c28d0b5/python-squid/src/main/java/org/sonar/python/PythonAstScanner.java#L100): Remove this unused method parameter "conf". [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1172)

File: python-squid/src/main/java/org/sonar/python/metrics/FileLinesVisitor.java [hunk 4]
```diff
@@ -45,37 +49,77 @@
...
+    linesOfComments = new HashSet<>();
+    seenFirstToken = false;
   }
 
+  /**
+   * Gets the lines of codes and lines of comments (with character #).
+   * Does not get the lines of docstrings. 
+   */
   @Override
   public void visitToken(Token token) {
```
Comment (SonarTech) [inline python-squid/src/main/java/org/sonar/python/metrics/FileLinesVisitor.java]: CRITICAL Refactor this method to reduce its Cognitive Complexity from 24 to the 15 allowed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3776)

File: python-squid/src/main/java/org/sonar/python/metrics/FileLinesVisitor.java [hunk 5]
```diff
@@ -45,37 +49,77 @@
...
-      if (trivia.isComment()) {
-        linesOfComments.add(trivia.getToken().getLine());
+    if (!ignoreHeaderComments || seenFirstToken) {
+      for (Trivia trivia : token.getTrivia()) {
+        if (trivia.isComment()) {
+          String[] commentLines = getContext().getCommentAnalyser().getContents(trivia.getToken().getOriginalValue())
+            .split("(\r)?\n|\r", -1);
+          int line = trivia.getToken().getLine();
+
+          for (String commentLine : commentLines) {
```
Comment (SonarTech) [inline python-squid/src/main/java/org/sonar/python/metrics/FileLinesVisitor.java]: CRITICAL Refactor this code to not nest more than 3 if/for/while/switch/try statements. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS134)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* CRITICAL 2 critical

Watch the comments in this conversation to review them.

File: python-squid/src/main/java/org/sonar/python/metrics/FileLinesVisitor.java [hunk 6]
```diff
@@ -45,37 +48,94 @@
...
+      }
+    }
+  }
+
+  /**
+   * Gets the lines of codes and lines of comments (with character #).
+   * Does not get the lines of docstrings.
+   */
+  @Override
   public void visitToken(Token token) {
```
Comment (SonarTech) [inline python-squid/src/main/java/org/sonar/python/metrics/FileLinesVisitor.java]: CRITICAL Refactor this method to reduce its Cognitive Complexity from 24 to the 15 allowed. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS3776)

File: python-squid/src/main/java/org/sonar/python/metrics/FileLinesVisitor.java [hunk 7]
```diff
@@ -45,37 +48,94 @@
...
+  private Set<Integer> linesOfDocstring = Sets.newHashSet();
   private final FileSystem fileSystem;
   private final Map<InputFile, Set<Integer>> allLinesOfCode;
 
-  public FileLinesVisitor(FileLinesContextFactory fileLinesContextFactory, FileSystem fileSystem, Map<InputFile, Set<Integer>> linesOfCode) {
+  public FileLinesVisitor(
+      FileLinesContextFactory fileLinesContextFactory,
+      FileSystem fileSystem,
+      Map<InputFile, Set<Integer>> linesOfCode,
+      boolean enableNoSonar,
```
Comment (ivandalbosco) [inline python-squid/src/main/java/org/sonar/python/metrics/FileLinesVisitor.java]: @pynicolas You're right. Removed. Now equivalent to enableNoSonar always set to true.

File: sonar-python-plugin/src/test/resources/org/sonar/plugins/python/squid-sensor/file1.py [hunk 8]
```diff
@@ -1,18 +1,9 @@
 # Integers
 00000
-1111L
```
Comment (ivandalbosco) [inline sonar-python-plugin/src/test/resources/org/sonar/plugins/python/squid-sensor/file1.py]: @pynicolas Sorry, restored.

File: python-squid/src/test/resources/docstring.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
+"""
+This is a module docstring
+"""
+
+def function_with_docstring():
+	"""This is a function docstring"""
+	pass
+
+def function_without_docstring():
+#   ^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (ivandalbosco) [inline python-squid/src/test/resources/docstring.py]: @pynicolas It was a blind copy from the test file for a Check class. Removed. Thanks!

File: python-squid/src/test/resources/docstring.py [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+	pass
+
+def function_with_empty_docstring():
+#   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+	""" """
+	pass
+
+def one_line_function_with_docstring(): "This is a function docstring"; pass
```
Comment (ivandalbosco) [inline python-squid/src/test/resources/docstring.py]: @pynicolas Agreed x 2. I had noticed this glitch but I left it aside as the Python way is to use separate lines for docstrings.

**Pull request merged**